src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -2548,6 +2647,326 @@ public final void addBatch() throws SQLServerException {
+            
+            // keep checking if it's escaped
+            while (localUserSQL.charAt(tempint + 1) == ']') {
+                localUserSQL = localUserSQL.substring(0, tempint) + localUserSQL.substring(tempint + 1);
+                tempint = localUserSQL.indexOf("]", tempint + 1);
+            }
+            
+            // we've found a ] that is actually trying to close the square bracket.
+            // If it's followed by a dot, then it's a database.
+            // Otherwise, it's the table.
+            if (localUserSQL.charAt(tempint + 1) == '.') {
+                String tempstr = localUserSQL.substring(1, tempint);
+                localUserSQL = localUserSQL.substring(tempint + 2);
+                // assume that "INSERT" and "INTO" has been found, since we're at the table part already.
+                return tempstr + "." + parseUserSQLForTableNameDW(true, true);
+            } else {
+                // return tablename
+                String tempstr = localUserSQL.substring(1, tempint);
+                localUserSQL = localUserSQL.substring(tempint + 1);
+                return tempstr;
+            }
+        }
+        
+        // do the same for ""
+        if (localUserSQL.substring(0, 1).equalsIgnoreCase("\"")) {
+            int tempint = localUserSQL.indexOf("\"", 1);
+            
+            // keep checking if it's escaped
+            while (localUserSQL.charAt(tempint + 1) == '\"') {
+                localUserSQL = localUserSQL.substring(0, tempint) + localUserSQL.substring(tempint + 1);
+                tempint = localUserSQL.indexOf("\"", tempint + 1);
+            }
+            
+            // we've found a " that is actually trying to close the quote.
+            // If it's followed by a dot, then it's a database.
+            // Otherwise, it's the table.
+            if (localUserSQL.charAt(tempint + 1) == '.') {
+                String tempstr = localUserSQL.substring(1, tempint);
+                localUserSQL = localUserSQL.substring(tempint + 2);
+                return tempstr + "." + parseUserSQLForTableNameDW(true, true);
+            } else {
+                // return tablename
+                String tempstr = localUserSQL.substring(1, tempint);
+                localUserSQL = localUserSQL.substring(tempint + 1);
+                return tempstr;
+            }
+        }
+        
+        // At this point, the next chunk of string is the table name (could have database name), without starting with [ or ".
+        StringBuilder sb = new StringBuilder();
+        while (localUserSQL.length() > 0) {
+            if (localUserSQL.charAt(0) == '.' || localUserSQL.charAt(0) == ' ' || localUserSQL.charAt(0) == '\t'
+                    || localUserSQL.charAt(0) == '(') {
+                if (localUserSQL.charAt(0) == '.') {
+                    localUserSQL = localUserSQL.substring(1);
+                    return sb.toString() + "." + parseUserSQLForTableNameDW(true, true);
```

Comment:
What if table name contains a `.`?

Reply:
Same whitespace check failure will happen here too.

Reply:
Table name containing . is handles. I made changes for the whitespace check.